tutorial/markdown/python/gemini-langchain-pdf-chat/gemini-langchain-pdf-chat-py.md

```diff
@@ -0,0 +1,579 @@
+
+```python
+template = """You are a helpful bot. If you cannot answer based on the context provided, respond with a generic answer. Answer the question as truthfully as possible using the context below:
+{context}
+
+Question: {question}"""
+
+prompt = ChatPromptTemplate.from_template(template)
+```
+
+### LLM Chain
+
+Large Language Models (LLMs) are a core component of LangChain. LangChain does not serve its own LLMs, but rather provides a standard interface for interacting with many LLMs. To be specific, this interface is one that takes as input a string and returns a string. We will use [Google AI](https://python.langchain.com/docs/integrations/llms/google_ai/) LLM Model with gemini model. We can also set other parameters like model, API_KEY, temperature to be used for this model.
+
+```python
+llm = GoogleGenerativeAI(
+    temperature=0.1,
+    model="models/gemini-1.5-pro",
+)
+```
+
+### Combining to a single chain
+
+We can combine these different modules to a single chain which will run synchronously. The pipe operator (|) is used to connect these components, forming a chain of operations.
+
+The input dictionary with the "context" and "question" keys is passed through each component. The retriever component retrieves relevant information from the vector store, which is then combined with the [question](https://python.langchain.com/docs/expression_language/primitives/passthrough/) to generate a prompt. This prompt is passed to the language model (llm), which generates a response based on both the question and the retrieved context. Finally, the output is parsed into a string by [StrOutputParser()](https://python.langchain.com/docs/modules/model_io/concepts/#stroutputparser).
+
+```python
+# RAG chain
+chain = (
+    {"context": retriever, "question": RunnablePassthrough()}
+    | prompt
+    | llm
+    | StrOutputParser()
+)
+```
+
+### Chain without RAG
+
+We will repeat the same process as above however this will not have the context from the vector store. Basically we will directly call LLM from the user question. Basically, every step is same just that we will not use retriever.
+
+```python
+template_without_rag = """You are a helpful bot. Answer the question as truthfully as possible.
+
+Question: {question}"""
+
+prompt_without_rag = ChatPromptTemplate.from_template(template_without_rag)
+
+llm_without_rag = GoogleGenerativeAI(
+    temperature=0,
+    model="models/gemini-1.5-pro",
+)
+
+chain_without_rag = (
+    {"question": RunnablePassthrough()}
+    | prompt_without_rag
+    | llm_without_rag
+    | StrOutputParser()
+)
+```
+
+### User Asks A Question
+
+This section creates an interactive chat interface where users can ask questions based on the uploaded PDF. The key steps are:
+
+1. Display a chat input box with the prompt "Ask a question based on the PDF".
+   When the user submits a question:
```

Comment:
Something is off here in terms of the formatting. Maybe a new line?

Reply:
Updated